# <center>The Output Files</center>

<center>Dr. W.J.B. Mattingly</center>

<center>Smithsonian Data Science Lab and United States Holocaust Memorial Museum</center>

<center>August 2021</center>

## Covered in this Chapter

1) The .tokens file<br>
2) The .entities file<br>
3) The .quote file<br>
4) The .supersense file<br>
5) The .book file<br>
6) The .book.html file<br>

## Introduction

In the last chapter, we looked at how to create a BookNLP pipeline and process a book or longer document. The goal of that process was to generate a collection of files within an output directory. In our case, we stored our files in "data/harry_potter". Within this repo, you will be able to examine the output files, but rather than making you switch between this textbook and the repo, I thought I would present the files in this chapter with screenshots. With that said, let's begin.

## The .tokens file

The very first file that we should analyze is the .tokens file. Essentially, this is a tab separated value file (TSV). A good way to think about a TSV is as a CSV where tabs are used to separate tabular data, rather than commas. Essentially, this is a dataset that can be viewed and analyzed in Excel. The very first line of the file will look something like this:

paragraph_ID	sentence_ID	token_ID_within_sentence	token_ID_within_document	word	lemma	byte_onset	byte_offset	POS_tag	fine_POS_tag	dependency_relation	syntactic_head_ID	event

As this can be a bit difficult to parse, I am going to load it up as a TSV file through Pandas so we can analyze it a bit better.

In [3]:
import pandas as pd

df = pd.read_csv("data/harry_potter/harry_potter.tokens", delimiter="\t")
df

paragraph_ID  sentence_ID  token_ID_within_sentence  \
0                 0            0                         0   
1                 0            0                         1   
2                 0            0                         2   
3                 0            0                         3   
4                 0            0                         4   
...             ...          ...                       ...   
99396          3030         6972                        12   
99397          3030         6972                        13   
99398          3030         6972                        14   
99399          3031         6973                         0   
99400          3031         6973                         1   

       token_ID_within_document      word     lemma  byte_onset byte_offset  \
0                             0     Harry     Harry           0           5   
1                             1    Potter    Potter           6          12   
2                             2       and       and          13          16   
3                             3       the       the          17          20   
4                             4  Sorcerer  Sorcerer          21          29   
...                         ...       ...       ...         ...         ...   
99396                     99396    summer    summer      439721      439727   
99397                     99397      ....      ....      439727      439731   
99398                     99398        \t    439731      439732       PUNCT   
99399                     99399       THE       the      439734      439737   
99400                     99400       END       END      439738      439741   

      POS_tag fine_POS_tag dependency_relation syntactic_head_ID event  
0       PROPN          NNP            compound                 1     O  
1       PROPN          NNP               nsubj                12     O  
2       CCONJ           CC                  cc                 1     O  
3         DET           DT                 det                 4     O  
4       PROPN          NNP                poss                 8     O  
...       ...          ...                 ...               ...   ...  
99396    NOUN           NN            npadvmod             99388     O  
99397   PUNCT            .               punct             99386     O  
99398      ''        punct               99386                 O   NaN  
99399     DET           DT                 det             99400     O  
99400   PROPN          NNP               nsubj             99398     O  

[99401 rows x 13 columns]

If you don't know what the block of code above does, please do not be concerned. We will not be dealing with Pandas in this textbook. If you are interested in Pandas, though, I have a free textbook on it entitled <a href="pandas.pythonhumanities.com">Introduction to Pandas</a>.

As you can see from the output above, we have something that looks like Excel, or tabular data. Let's break this down a bit and explain what each column represents:

- paragraph_ID = the index of the paragraph, starting at paragraph 1 being 0 and moving up to 3031 in our case.
- sentence_ID = same as the paragraph_ID, but with sentences
- token_ID_within_sentence = same as a the two above, but with a token count by sentence, resetting with each sentence.
- token_ID_within_document = same as above, but where tokens keep going up in value throughout the whole document, starting at 0 and ending, in our case, at 99400.
- word = this is the raw text of the word
- lemma = this is the root of the word
- byte_onset = think of this as the start character index
- byte_offset - think of this as the concluding character index
- POS_tag = the Part of Speech (based on spaCy)
- fine_POS_tag = a more granular understanding of the Part of Speech
- dependency_relation = this is equivalent to spaCy's dep tag.
- syntactic_head_ID = This points to the head of the current token so that you can understand how a token relates to other words in the sentence
- event = this tells you if the toke

## The .entities File

## The .quote File

## The .supersense file

## The .book File

## The .book.html File